pkg/resource/inference_component/hooks.go

```diff
@@ -0,0 +1,166 @@
+// proceed with UpdateInferenceComponent call.
+// Update is blocked in the following cases:
+//  1. while InferenceComponentStatus != InService (handled by requeueUntilCanModify method).
+//  2. InferenceComponentStatus == Failed.
+//  3. A previous update to the InferenceComponent with same spec failed.
+//
+// Method returns nil if InferenceComponent can be updated,
+// otherwise InferenceComponentUpdateError depending on above cases.
+func (rm *resourceManager) customUpdateInferenceComponentPreChecks(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) error {
+	latestStatus := latest.ko.Status.InferenceComponentStatus
+	if latestStatus == nil {
+		return nil
+	}
+
+	failureReason := latest.ko.Status.FailureReason
+
+	desiredSpec := desired.ko.Spec.DeepCopy()
+	desiredSpec.Tags = nil
+
+	var lastSpecForUpdateString *string = nil
+	// get last endpoint config name used for update from annotations
+	annotations := desired.ko.ObjectMeta.GetAnnotations()
+	for k, v := range annotations {
+		if k == lastSpecForUpdateAnnotation {
+			lastSpecForUpdateString = &v
+		}
+	}
+
+	var lastSpecForUpdate *svcapitypes.InferenceComponentSpec
+
+	if lastSpecForUpdateString != nil {
+		err := json.Unmarshal([]byte(*lastSpecForUpdateString), &lastSpecForUpdate)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Case 2 - InferenceComponentStatus == Failed
+	if *latestStatus == svcsdk.InferenceComponentStatusFailed ||
+		// Case 3 - A previous update to the InferenceComponent with same spec failed
+		// Following checks indicate FailureReason is related to a failed update
+		(failureReason != nil && lastSpecForUpdateString != nil &&
+			EqualInferenceComponentSpec(desiredSpec, lastSpecForUpdate)) {
+		// 1. FailureReason alone doesn't mean an update failed it can appear because of other
+		// reasons(patching/scaling failed).
+		// 2. desiredSpec == lastSpecForUpdate only tells us an update was tried with lastSpecForUpdate
+		// but does not tell us anything if the update was successful or not in the past because
+		// it is set if updateInferenceComponent returns 200 (async operation).
+		// 3. Now, sdkUpdate can execute because of change in any field in Spec.
+
+		// 1 & 2 does not guarantee an update Failed. Hence, we need to look at `lastSpecForUpdate` to determine if the update was unsuccessful
+		// `desiredSpec != latestSpec` + `desiredSpec == lastSpecForUpdate
+		//`+ `FailureReason != nil` indicate that an update is needed, has already been tried and failed.
+		return awserr.New("InferenceComponentUpdateError", "Unable to update inference component."+
+			" Check FailureReason.", nil)
+	}
+
+	return nil
+}
+
+// EqualInferenceComponentSpec checks if two InferenceComponentSpec instances are equal
+func EqualInferenceComponentSpec(desiredSpec *svcapitypes.InferenceComponentSpec,
+	lastSpec *svcapitypes.InferenceComponentSpec) bool {
+	if desiredSpec == nil || lastSpec == nil {
+		return desiredSpec == lastSpec
```

Comment:
is this because .DeepEqual cannot handle nil?

Reply:
Yes, Reflect package does not handle nil pointers gracefully. 

The reason for the nil check before using reflect.DeepEqual is to prevent a potential panic caused by attempting to dereference a nil pointer. If either desiredSpec or lastSpec is nil, attempting to pass them to reflect.DeepEqual directly would result in a runtime panic because reflect.DeepEqual expects both of its arguments to be non-nil.